tbd/handler/handler.go

```diff
@@ -0,0 +1,72 @@
+package handler
+
+import (
+	"fmt"
+	"net/http"
+)
+
+type Logger interface {
+	Request(*http.Request, error)
+}
+
+type Handler func(w http.ResponseWriter, r *http.Request) error
+
+type ErrorFunc func(w http.ResponseWriter, r *http.Request, statusCode int, err error)
+
+func New(redirectBaseURL, authURL string, errorFunc ErrorFunc) func(Handler) http.Handler {
+	return func(next Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			if err := next(w, r); err != nil {
+				statusCode := http.StatusInternalServerError
+
+				switch v := err.(type) {
+				case RedirectError:
+					http.Redirect(w, r, redirectBaseURL+v.RedirectTo(), http.StatusFound)
+					return
+				case StatusError:
+					if v.StatusCode() == http.StatusUnauthorized {
+						http.Redirect(w, r, authURL, http.StatusFound)
+						return
+					}
+
+					if v.StatusCode() == http.StatusForbidden || v.StatusCode() == http.StatusNotFound {
+						statusCode = v.StatusCode()
+					}
+				}
+
+				w.WriteHeader(statusCode)
+				errorFunc(w, r, statusCode, err)
```

Comment:
Instead of this I could have hard-coded the implementation used here, where we log and print an error template, but I felt like that might be too restrictive?